From 7565cab79e5c50be73a77fe2f8a345dc4e2688ab Mon Sep 17 00:00:00 2001
From: Hansong Zhang <hsz@google.com>
Date: Wed, 8 Aug 2018 11:31:28 -0700
Subject: [PATCH 4/8] Check remaining frame length in rfc_process_mx_message

Bug: 111936792
Bug: 80432928
Test: manual
Change-Id: Ie2c09f3d598fb230ce060c9043f5a88c241cdd79
(cherry picked from commit 0471355c8b035aaa2ce07a33eecad60ad49c5ad0)
---
 stack/rfcomm/rfc_ts_frames.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/stack/rfcomm/rfc_ts_frames.c b/stack/rfcomm/rfc_ts_frames.c
index 4690200..3f4f512 100644
--- a/stack/rfcomm/rfc_ts_frames.c
+++ b/stack/rfcomm/rfc_ts_frames.c
@@ -28,6 +28,7 @@
 #include "l2c_api.h"
 #include "port_int.h"
 #include "rfc_int.h"
+#include <log/log.h>
 
 /*******************************************************************************
 **
@@ -702,6 +703,14 @@ void rfc_process_mx_message (tRFC_MCB *p_mcb, BT_HDR *p_buf)
     UINT8        ea, cr, mx_len;
     BOOLEAN      is_command;
 
+    if (length < 2)
+    {
+        RFCOMM_TRACE_ERROR2("%s: Illegal MX Frame len when reading EA, C/R. len:%d < 2",
+                            __func__, length);
+        android_errorWriteLog(0x534e4554, "111937065");
+        GKI_freebuf (p_buf);
+        return;
+    }
     p_rx_frame->ea   = *p_data & RFCOMM_EA;
     p_rx_frame->cr   = (*p_data & RFCOMM_CR_MASK) >> RFCOMM_SHIFT_CR;
     p_rx_frame->type = *p_data++ & ~(RFCOMM_CR_MASK | RFCOMM_EA_MASK);
@@ -724,6 +733,14 @@ void rfc_process_mx_message (tRFC_MCB *p_mcb, BT_HDR *p_buf)
 
     if (!ea)
     {
+        if (length < 1)
+        {
+            RFCOMM_TRACE_ERROR2("%s: Illegal MX Frame when EA = 0. len:%d < 1",
+                                __func__, length);
+            android_errorWriteLog(0x534e4554, "111937065");
+            GKI_freebuf (p_buf);
+            return;
+        }
         mx_len += *p_data++ << RFCOMM_SHIFT_LENGTH2;
         length --;
     }
@@ -800,7 +817,14 @@ void rfc_process_mx_message (tRFC_MCB *p_mcb, BT_HDR *p_buf)
         return;
 
     case RFCOMM_MX_MSC:
-
+        if (length != RFCOMM_MX_MSC_LEN_WITH_BREAK &&
+            length != RFCOMM_MX_MSC_LEN_NO_BREAK)
+        {
+            RFCOMM_TRACE_ERROR2("%s: Illegal MX MSC Frame len:%d", __func__, length);
+            //android_errorWriteLog(0x534e4554, "111937065");
+            GKI_freebuf (p_buf);
+            return;
+        }
         ea                   = *p_data & RFCOMM_EA;
         cr                   = (*p_data & RFCOMM_CR_MASK) >> RFCOMM_SHIFT_CR;
         p_rx_frame->dlci = *p_data++ >> RFCOMM_SHIFT_DLCI;
-- 
2.7.4

